src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs

```diff
@@ -376,10 +376,22 @@ private static Block ConvertToMarkupAttributeBlock(Block block, bool isBoundNonS
             return blockBuilder.Build();
         }
 
-        private static Block RebuildChunkGenerators(Block block)
+        private static Block RebuildChunkGenerators(Block block, bool isBound)
         {
             var builder = new BlockBuilder(block);
 
+            // Don't want to rebuild unbound dynamic attributes. They need to run through the conditional attribute
+            // removal system at code generation time.
+            if (!isBound &&
+                block.Children.Any(
+                    child => child.IsBlock &&
+                    ((Block)child).ChunkGenerator is DynamicAttributeBlockChunkGenerator))
```

Comment:
Unclear how this condition relates to the block just below which special-cases `builder.ChunkGenerator is DynamicAttributeBlockChunkGenerator` and sets `builder.ChunkGenerator` in the same way.

All-up, probably better to provide a description of what we _do_ want. What are the cases where `builder.ChunkGenerator` must be `ParentChunkGenerator.Null`? When must children be rebuilt and many `Span`s changed to use a `MarkupChunkGenerator`?